libs/contrib/Data/Nat/Factor.idr

```diff
@@ -0,0 +1,332 @@
+module Data.Nat.Factor
+
+import Data.Fin
+import Data.Fin.Extra
+import Data.Nat
+import Decidable.Order
+import Syntax.PreorderReasoning
+
+%default total
+%access public export
+
+
+||| Factor n p is a witness that p is indeed a factor of n,
+||| i.e. there exists a q such that p * q = n.
+data Factor : Nat -> Nat -> Type where
```

Comment:
This is maybe getting into bikeshedding territory, but what about
changing this from `Factor n p` to `Factor p n`? `Factor n p` reads as
"n has p as a factor", but I feel like the more natural way to say it
would be "p is a factor of n", or alternatively, "p evenly divides n".
With infix syntax, it could even be `p | n` or something like that.
Similar with `CommonFactor`. Instead of saying "a and b both have p as
a factor", the reading would be "p is a factor of both a and b" or "p
divides both a and b".

Besides being more natural (IMO), this would clarify the way in which
`CommonFactor` is symmetric. Given some "p", the relation is "has
factor p in common with", as in "a has factor p in common with b".
Something like this:

```idris
using (p : Nat)
  Preorder Nat (CommonFactor p) where

  Equivalence Nat (CommonFactor p) where
```

Reply:
Unfortunately, we can't have `(|)` as a type constructor, since it is already reserved as a part of Idris syntax. Nontheless, this is probably a better choice, so I'll change it.